Add Google Cloud Storage backend using the gcloud-python library

scjody opened the pull request:
Based on https://github.com/jschneier/django-storages/pull/146

This uses the native `gcloud-python` library, which is recommended by Google for production setups and allows finer-grained authentication than current solutions (which only work when Google Cloud Storage is used in S3 compatibility mode).

* [x] Review code (see: https://github.com/jschneier/django-storages/pull/146#issuecomment-266873319)
* [x] Add tests
* [x] Add documentation

scjody commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,241 @@
+            self._file = None
+
+
+class GoogleCloudStorage(Storage):
```
@eirsyl The existing backends all use the `@deconstructible` decorator here but I'm not sure why. Commit comments suggest it has something to do with migrations but I'm able to migrate to a `FileField` using this backend for storage with this code as is. Therefore I'm inclined to leave the decorator out and if someone needs it, they can add it. What do you think?

scjody replied:
Never mind, I figured out why the decorator is needed and added it.

scjody commented:
I reviewed the code and made some minor changes. It looks good - very close to the S3 implementation, but S3 and Google Cloud Storage are very close to begin with.

Onto tests!

scjody commented:
@eirsyl @jschneier Does this look reasonable?

Next steps on our end are that we'll be using this in our staging environment within the next 2 weeks.

eirsyl commented:
Great work! Looks good to me 👍

scjody commented:
@jschneier We've been using this successfully in our staging environment for several days. Would you be able to merge it soon please, or let me know what's still missing?

scjody commented:
@jschneier Can you please take a look at this? The README says to bug you until it's merged, so that's what I'm doing 😸.   It also said to add myself to AUTHORS so I added myself and @eirsyl. @eirsyl please let me know if you want a different name or don't want to be credited.

scjody commented:
@jschneier Will you have time to look at this sometime soon please?

@ZuluPro Is this something you could look at instead? (I can't tell if you have write access but it looks like you might.)

jschneier commented:
Yes, currently on vacation but have set aside Monday afternoon to do maintenance on this library, thanks for your patience.

> On Feb 10, 2017, at 14:36, Jody McIntyre <notifications@github.com> wrote:
> 
> @jschneier Will you have time to look at this sometime soon please?
> 
> @ZuluPro Is this something you could look at instead? (I can't tell if you have write access but it looks like you might.)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

sherbondy commented:
I think there might be a renewed interest in getting this PR merged given the events that unfolded today :/

scjody commented:
@jschneier Just a reminder: can you please review this soon? I've updated AUTHORS again to fix the merge conflict.

p0r commented:
This is great development, and I hope it gets merged soon. I'd like to present one wish though - please take #281 into account and make use of tempfile.SpooledTemporaryFile optional / configurable.

stinky2nine commented:
@jschneier Friendly ping.

mhindery commented:
Would also love this to be merged, as the current solution for GCS is far from ideal ...

jschneier commented:
Hi.

Can someone articulate what the issue is with the current `gs.py` binding? I'm happy to merge something like this since it seems that there is demand for it. I also greatly appreciate that the code is similar to the other backends that exist although I can't help but feel we can probably find a better abstraction since there is quite a bit shared (and that the code quality of `S3BotoStorage` is showing its age a bit).

When we do merge this I'd like to deprecate `gs.py`.

stinky2nine commented:
@jschneier `gs.py` is using Google Cloud Storage's [XML Interoperable API](https://cloud.google.com/storage/docs/interoperability) which uses keyed-hash message authentication code (a.k.a. developer keys) that is linked to your Google account. Interoperable API is really meant for migration to Google Cloud Storage. The biggest problem with the developer keys is security and privacy. Developer keys should not be shared with anyone as they can be used to gain access to other Google Cloud Storage buckets linked to your Google account.

I believe that this pull request depends on [`gcloud.storage.client.Client`](https://googlecloudplatform.github.io/google-cloud-python/0.18.1/storage-client.html) which would require using [OAuth 2.0](https://cloud.google.com/storage/docs/authentication) for Google Cloud Storage API authN and authZ.

mhindery commented:
As extra note on the explanation above:

- If you work with multiple developers on a project, they each have to enable the interoperability mode and create keys which then have to be picked up by you project config/settings. This means quite some unnecessary work and config. Alternatively you would have to distribute the credentials for 1 account to everybody. This is not wanted.
- When deploying on Google Cloud instances or (kubernetes) clusters, there are by default application credentials present for authentication against GCS (service accounts). It is these services accounts that are used to manage roles and authentication on the Google Cloud platform everywhere. On developer machines, they are also present system-wide for all applications to pick up by default. Each program thus runs under the dev account when running locally, and uses the instance/cluster's account when deployed somewhere, and this all happens automatically. They are the recommended way of authenticating and are used throughout the platform (https://developers.google.com/identity/protocols/application-default-credentials), available for multiple languages.

scjody commented:
@jschneier It's mostly about authentication as others have articulated above. (I've personally tested the new functionality using an instance's service account and with a developer's user account, but I expect any authentication type supported by the `google-cloud-python` library will also work.)

I also believe that since the new backend uses the official Google library directly and does not require a compatibility mode intended only for short term data migrations, it is more likely to be reliable in the long term.

scjody commented:
@mhindery Are you sure you're in the same virtualenv in both cases? Can you try debugging the `try` part of the code above using `ipdb`? Or removing the `try`/`except` to see what the exact error is?

scjody commented:
@jschneier I just rebased this to deal with your recent merges. This is ready to go again; please let me know if you have any concerns or want any changes.

There are some strange things in the existing backends, especially s3boto, but addressing that is beyond the scope of this PR (and in many cases would result in backwards-incompatible changes). I did move some code from `s3boto` to `utils` to avoid duplication between the backends.

jschneier commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,226 @@
+    def _save(self, name, content):
+        cleaned_name = self._clean_name(name)
+        name = self._normalize_name(cleaned_name)
+        size = getattr(content, 'size')
```
Why do this with `getattr`? If it doesn't exist this is going to fail with `AttributeError` anyway so it can just be inlined right i.e. `size=content.size`. Does the `.upload_from_file` not automatically infer size as well?

scjody replied:
I'm not sure why it was done this way anymore, but your suggestion makes sense so I'll implement it.

`upload_from_file()` can infer the size if it's an actual file, but not if it's a `StringIO` or similar.

jschneier commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,226 @@
+    client_class = Client
+    file_class = GoogleCloudFile
+
+    not_found_exception = NotFound
```
I think we can remove this and probably some of the other class based properties. In this case I think it is there to make it easy to create the `gs.py` file unless you can think of a solid reason to keep it.

scjody replied:
OK, I'll remove this and anything like it.

jschneier commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,226 @@
+    def modified_time(self, name):
+        name = self._normalize_name(self._clean_name(name))
+        blob = self.bucket.get_blob(self._encode_name(name))
+        return blob.updated if blob else None
```
Need to add `get_modified_time` as well. Is this a naive time? This has to be naive while `get_modified_time` depends on the setting `USE_TZ`. Looking at `S3BotoStorage` might be useful.

scjody replied:
This is currently aware, so I'll make it naive and add the other method.

jschneier commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,226 @@
+                                       "setting GS_AUTO_CREATE_BUCKET to "
+                                       "``True``." % name)
+
+    def _clean_name(self, name):
```
Should we still have this method?

scjody replied:
It doesn't need to be a method, no. I've changed it to just call `clean_name` directly.

jschneier commented on tests/test_gcloud.py:
```diff
@@ -0,0 +1,210 @@
+        self.storage._bucket.get_blob.return_value = None
+        self.storage.file_overwrite = False
+        self.assertEqual(self.storage.get_available_name(self.filename), self.filename)
+        self.storage._bucket.get_blob.assert_called_with(self.filename)
```
Let's add a unicode test name.  Would like to not have Python 2.7 issues aka [this bug](https://github.com/jschneier/django-storages/pull/217).

scjody replied:
Good idea, will do.

jschneier commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,226 @@
+        name = self._normalize_name(self._clean_name(name))
+        file_object = self.file_class(name, mode, self)
+        if not file_object.blob:
+            raise IOError('File does not exist: %s' % name)
```
How about we stick with `NotFound` throughout.

scjody replied:
s3boto raises `IOError` in this case, so I want to stick with that to avoid breaking things for someone who switches from s3boto to gcloud.

jschneier replied:
Sounds good :+1:

jschneier commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,226 @@
+
+    def size(self, name):
+        name = self._normalize_name(self._clean_name(name))
+        blob = self.bucket.get_blob(self._encode_name(name))
```
Ugh really do not like that `get_blob` suppresses errors, it should not do that. I'd prefer to fail loudly when we do an action on something that doesn't exist. Size of 0 doesn't make too much sense to me. Maybe we wrap it up in our own `_get_blob` and just `raise NotFound` when it's...not found? Thoughts?

scjody replied:
Yes, that makes sense. I made that change here and in `modified_time` and `url`.

jschneier commented:
@scjody okay, left a few thoughts, should have done it as a review, sorry about that. going to let it stew for a bit and then maybe add some other thoughts. let me know what you think about them.

also, when we take this in you are okay if i give you access and you can be a maintainer of this particular piece of code?

jschneier commented:
Ah, can you also add a deprecation warning for `gs.py`! Forgot about that. And include the authentication/security reasons from above in this thread.

dummerbd commented:
It looks like @scjody already has a good handle on this, but if you need any help I can lend a hand. I just spent most of today writing up my own backend using the `google.cloud` library before I came across this PR, so I might be able to help out.

mhindery commented:
@scjody I have looked into the import error. I am certainly in the right environment (see attached traceback; print the installed packages, check my gcloud credentials, print django storage class -> error)
[traceback.txt](https://github.com/jschneier/django-storages/files/913318/traceback.txt)

I then tried the import with `ipdb` which gives 
```python
In [1]: from django.core.files.storage import default_storage

In [2]: print default_storage.__class__
> /Users/mhindery/repositories/otainsight/parity/storages/backends/gcloud.py(13)<module>()
     11     import ipdb
     12     ipdb.set_trace(context=5)
---> 13     from google.cloud.storage.client import Client
     14     from google.cloud.storage.blob import Blob
     15     from google.cloud.exceptions import NotFound

ipdb> n
ImportError: 'No module named compat'
> /Users/mhindery/repositories/otainsight/parity/storages/backends/gcloud.py(13)<module>()
     11     import ipdb
```

Some googling did not lead me to a solution on this, neither did I find something on it in the issues of the google-cloud-python repo.

jschneier commented:
@mhindery there used to be a `compat.py` file in this project but it has since been removed. You might need a rebased version of this branch.

jschneier commented on storages/backends/gcloud.py:
```diff
@@ -203,24 +191,33 @@ def listdir(self, name):
+        blob = self.bucket.get_blob(name)
+
+        if blob is None:
+            raise NotFound('File does not exist')
```
Can you include the `name` in the error message.

scjody replied:
👍

jschneier commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,223 @@
+
+
+class GoogleCloudFile(File):
+    def __init__(self, name, mode, storage, buffer_size=None):
```
`buffer_size` is unused?

scjody replied:
Yes, well spotted!

jschneier commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,223 @@
+
+    file = property(_get_file, _set_file)
+
+    def read(self, *args, **kwargs):
```
`read` should not take in arbitrary `args` and `kwargs`

scjody replied:
Fixed.

jschneier commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,223 @@
+            raise AttributeError("File was not opened in read mode.")
+        return super(GoogleCloudFile, self).read(*args, **kwargs)
+
+    def write(self, content, *args, **kwargs):
```
nor should `write`

scjody replied:
Fixed.

scjody commented:
@jschneier I addressed all your comments and added the deprecation warning. I think this is ready to go now.

> also, when we take this in you are okay if i give you access and you can be a maintainer of this particular piece of code?

Yes, no problem at all.

mhindery commented:
FWIW I have got it working on our staging environment (for some reason there was a storages/backends/google.py file leftover in my fork I used to try this out).

I am also wondering if it is possible to not set the ACL when uploading files / creating buckets. Instead, just user/inherit the default permissions we have configured in the web UI. It seems now it has to set it to something, which is by default `public-read`. We would like to configure this in the Gcloud UI, not in the django settings.

scjody commented:
> I am also wondering if it is possible to not set the ACL when uploading files / creating buckets. Instead, just user/inherit the default permissions we have configured in the web UI.

@mhindery That should work with the PR as is. I've been integration testing the PR with our software on a GCE server that authenticates to GCS using instance credentials and it works. Can you provide more detail on why you need to configure the ACL in the Django settings?

jschneier commented on storages/backends/gs.py:
```diff
@@ -1,3 +1,13 @@
+# Storage. The biggest problem with the developer keys is security and
+# privacy. Developer keys should not be shared with anyone as they can
+# be used to gain access to other Google Cloud Storage buckets linked
+# to your Google account.
```
The text is good but the user won't know. Usually I handle this the same as Django:

```python
import warnings
warnings.warn(
       "...text here", PendingDeprecationWarning)
```

Side question: I think that the xml compatibility layer will live on in the libcloud backend, does that seem correct?

scjody replied:
OK, I'll make it a warning.

I believe you're correct about libcloud, but I haven't tested it myself.

jschneier commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,233 @@
+    def url(self, name):
+        # Preserve the trailing slash after normalizing the path.
+        name = self._normalize_name(clean_name(name))
+        blob = self._get_blob(self._encode_name(name))
```
Let's push the `_encode_name(_normalize_name(clean_name(name)))` into `_get_blob`. It's always the same.

jschneier replied:
Sort of wondering if we can just do something like `_get_name(name)` and have that method alone do all 3 things, they basically always come together except in a couple places.

scjody replied:
Yes we could, but there are strange cases like the one you highlighted above. The way things are written now, it's easy to verify that `gcloud` works in the same way as `s3boto`.

I can make that change in this PR if you really want me to, but I think the right way to do it would be to merge this PR first then clean up this backend, `s3boto`, and any other backend that does this sort of thing in a separate PR.

jschneier replied:
Yeah, let's not touch it. We have all these backends I'm going to fix all this in all backends with common code in one go later.

jschneier commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,233 @@
+        cleaned_name = clean_name(name)
+        name = self._normalize_name(cleaned_name)
+
+        content.name = cleaned_name
```
Curious as to why `content.name` is not wrapped in `_encode_name`?

scjody replied:
This is for compatibility with `s3boto`. My basic approach is that someone should be able to take code that works with `s3boto`, drop in `gcloud`, and have it work (including quirks like this).

We do run `_encode_name` before creating the remote object, but the unencoded name is passed to `upload_from_file` (in `content`), similarly to how `s3boto` passes the unencoded name to `set_contents_from_file`.

stinky2nine commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,233 @@
+    credentials = setting('GS_CREDENTIALS', None)
+    bucket_name = setting('GS_BUCKET_NAME', None)
+    auto_create_bucket = setting('GS_AUTO_CREATE_BUCKET', False)
+    default_acl = setting('GS_DEFAULT_ACL', 'public-read')
```
@scjody I am wondering if the default here should be set to `None` and if `None`, ensure to use the default [object ACLs](https://cloud.google.com/storage/docs/access-control/lists#defaultobjects) set on the bucket.

scjody replied:
This is only used in the case where we create the bucket (`auto_create_bucket = True`). In that case we need a default because there are no default object ACLs we can use.

scjody replied:
I'll simplify the code and docs though. There's no need for two settings here.

stinky2nine commented on docs/backends/gcloud.rst:
```diff
@@ -0,0 +1,190 @@
+
+``GS_DEFAULT_ACL`` (optional)
+
+If set to ``private`` changes uploaded file's Access Control List from the default permission ``public-read`` to give owner full control and remove read access from everyone else. 
```
Perhaps list the list of the options here or reference to https://cloud.google.com/storage/docs/access-control/lists#predefined-acl.

In addition, the predefined ACL aliases for JSON API are different than XML API.

JSON API | XML API
--- | ---
private | private
bucketOwnerRead | bucket-owner-read
bucketOwnerFullControl | bucket-owner-full-control
projectPrivate | project-private
authenticatedRead | authenticated-read
publicRead | public-read
publicReadWrite | public-read-write

scjody replied:
From my understanding of the docs, you should be able to use either.

Unfortunately while testing this I ran into what's either a bug in the `google-cloud` python library or a huge misunderstanding on my part. I'm going to submit a support request to Google and see where that goes, and I'll update the PR when I have more information.

stinky2nine replied:
It seems like `google-cloud` is using JSON API and does **transform** predefined ACL aliases from XML to JSON format [here](https://github.com/GoogleCloudPlatform/google-cloud-python/blob/master/storage/google/cloud/storage/acl.py#L493). I would suggest only reference to JSON format here, e.g. `s/public-read/publicRead/g'.

scjody commented on storages/backends/gcloud.py:
```diff
@@ -0,0 +1,233 @@
+        if num_bytes is not None:
+            return super(GoogleCloudFile, self).read(num_bytes)
+        else:
+            return super(GoogleCloudFile, self).read()
```
Unfortunately we do. [Django defines the default argument to be `None`](https://docs.djangoproject.com/en/dev/ref/files/file/#django.core.files.File.read) but `SpooledTemporaryFile` (and Python `file` objects in general) can't handle the case where the first argument (`size`) is `None`:

```
Traceback (most recent call last):
  File "/home/ubuntu/gcs_integration/westore/tests.py", line 31, in test_write_to_storage
    self.assertEqual(f.read(), 'storage contents')
  File "/home/ubuntu/gcs_integration/storages/backends/gcloud.py", line 56, in read
    return super(GoogleCloudFile, self).read(num_bytes)
  File "/usr/lib/python2.7/tempfile.py", line 585, in read
    return self._file.read(*args)
TypeError: an integer is required
```

Instead the argument has to be omitted or set to a negative number.

The reason you haven't run into this in the `s3boto` backend is because [it uses `*args and **kwargs`](https://github.com/jschneier/django-storages/blob/e52a127523fdd5be50bb670ccad566c5d527f3d1/storages/backends/s3boto.py#L134-L137), which means that if the user omits the argument there's no `None` to pass through to the underlying `read()`. I believe if the user _does_ specify `num_bytes=None` s3boto will break, but I haven't tested this.

I agree that `num_bytes` is an unfortunate name 😿.

jschneier replied:
Interesting, that appears to be fixed on Python 3.5 at least.

jschneier replied:
I guess the only way to do it would be something like:
```python
if num_bytes is None:
  num_bytes = -1
...
```
Also I think we can just change the name from `num_bytes` to `size` and the default to `-1`, this is a new backend and I'll note it in the changelog if it is surprising.

jschneier replied:
To be clear I'm not usually very cavalier about breaking backwards compat but I think we'll have to go and do it once in a biggish way to wrestle all of these things to decent shape. We can leave this one here if you want for s3boto interaction for now until we hit that point but I'd be pretty surprised if anyone is kwarg'ing the arg to `read`.

scjody replied:
I'd be surprised too but [Django specifies `num_bytes` as the kwarg](https://docs.djangoproject.com/en/dev/ref/files/file/#django.core.files.File.read) so it would be reasonable for someone to do that, and really confusing to debug if they did.

Reading the Django docs a bit more closely I noticed that the text gives the name of the argument as `size`, so there's an inconsistency there. I think the best way forward is to submit a PR to fix the Django docs, wait for a version with the fix to be released, _then_ change this argument (here and in other backends).

In any case I've made the `if` change you suggested because your suggestion makes it a bit clearer.

jschneier replied:
Good call: https://github.com/django/django/pull/8379

scjody commented on storages/backends/gs.py:
```diff
@@ -1,3 +1,14 @@
+import warnings
+warnings.warn("DEPRECATION NOTICE: This backend is deprecated in favour of the "
```
@jschneier Like this, or would you like me to line wrap the warning output?

jschneier replied:
The syntax is fine. `warn` takes a second argument of the type to issue which in this case should be `DeprecationWarning` or `PendingDeprecationWarning`.

scjody replied:
I intended to do that but forgot. Thanks for noticing it!

scjody commented:
@jschneier I made the warning change you requested and commented on your other requests. Please don't merge this until I deal with https://github.com/jschneier/django-storages/pull/236#discussion_r111692934 though.

scjody commented:
@stinky2nine I addressed your comments by:
* Simplifying the ACL options to one (`GS_AUTO_CREATE_ACL`).
* Improving the documentation, including a link to the list on Google's site and a note on what I learned from my support request.

Thanks for the input!

@jschneier This is ready to go again. Please let me know if you want any more changes. I commented above re: the `*_name` methods, but I'm happy to make the change you suggested if you think that's the best way forward.

jschneier commented:
@scjody I just gave you collaboration access, fix up my last couple of comments at your digression and then go ahead and merge.

stinky2nine commented:
@scjody Looks good to me. I work at Google so feel free to ask any questions and thanks for putting together this pull request.

dummerbd commented:
Anyone got an ETA on when this version will be released?

Sgiath commented:
@scjody This is great I am able to use it properly but I have one problem: how I can make all objects uploaded by Django to the Google Storage publicly visible? Are there some settings for it? Maybe I can set it up on Google Cloud Storage itself? @stinky2nine 
Thanks
